### ISSUE TITLE
[class].default.[method] is not a function / initializer is undefined

### ISSUE THREAD

**Issue by rvan1** [ember-cli/ember-cli#3767]
My EAK 'app.js' creates an 'i18n' Ember Object, as a property of 'App', then in the 'beforeModel' method of 1st Route, it calls 'App.I18n.localeFor(localCode)', which is a method of the 'I18n' Object (an ../app/i18n.js file). Now that I'm migrating to Ember CLI, 'App is not defined', so I can't call any 'I18n' methods.

My 'app.js' makes a 'loadInitializers()' call (which might be a deprecated EAK practice) passing in 'App' and the 'modulePrefix', and so I can import the Object into my Route with "import I18n from ''ibe/i18n", but when I try to call its 'localeFor()' method, I get 'Ii18n.default.localeFor is not a function'. If I 'console.out' the Object, I see 'unknown mixin' and can't introspect it's properties.

Since the CLI-based Ember makes it super simple to setup the App either with utils, mixins or initializers, I've tried to inject this 'I18n' object into the application with an 'initializer', but with this I'm getting 'initializer is not defined'. 

The docs I see give a short description with no example. The examples I see reference 'App' as a global, which isn't really a valid approach anymore, or they don't 'export the initializer', which in my case results in an error if I don't, or they stub out a method with 'your code here' in the body. 

How can I get my custom object into the application and into the 1st Route 'beforeModel' call, so I can setup appropriate Localized translations like I do in the EAK version?

// app/initializers/initializer-i18n.js
```
export default Ember.Application.initializer({
  name: "initializer-i18n",
  initialize: function(container, application) {
    application.inject('route', 'initializer-i18n', 'i18n');
  }
});
```

**Comment by stefanpenner** [comment]
Yes, this has changed since your version of EAK I suppose, anyways as ember cli has generators for most things, I would suggest looking them for ideas on what something should be structured as. We try to keep those generators up to date with latest idioms.

something like:
```
ember g initializer my-cool-initializer
```

also for reference, the output you want is likely as follows:

```js
export default {
  name: "initializer-i18n",
  initialize: function(container, application) {
    application.inject('route', 'initializer-i18n', 'i18n');
  }
);
```

**Comment by rvan1** [comment]
Not sure what has changed, global reference of 'App' I guess. The EAK version doesn't use an initializer this way, but 'app.js' does call 'loadInitiializers', as does the only working version of a CLI-based app I have, the 'cli-todo-mvc' example, so I'm thinking to keep that.  With your code I get a build error until I change the last ) to }, then with this I get 'invalid fullname, expected: 'type:name' got 'i18n'.

Is there some simple, straight-forward approach to getting an object set onto the 'application', then into a route 'beforeModel' call? It seems like that should be an easy task.

**Comment by stefanpenner** [comment]
i have updated my above example slightly.

This doesn't appear to be an ember-cli bug, rather some questions. Let me recommend stack overflow or the ember forms as a better place, as those are much better venues for non-bug related Q/A. 

Feel free to cross link those questions here, but I will be closing this issue.

**Comment by rvan1** [comment]
I tried with the generator and can see I need to export the function AND export the 'default { body }'. Or maybe either syntax works. The docs imply a convention of 'initializer-name', but you suggest 'initializer' at the end so seems it's not really a convention, so then becomes hard to tell what's a convention and what isn't. Now I also get the 'expected type:name' error, but my class doesn't have a 'type'. I thought 'type' was just a sub directory.

**Comment by rvan1** [comment]
Got it working now. 

export default {
  name: "i18n-init",
  initialize: function(container, application) {
    application.inject('route', 'i18n-init', 'i18n:main');
  }
};

Appears the left side (type) needs to be my object 'class' (I18n but referenced as i18n), and the right side I set as 'main' since needed to be some valid keyword but not sure why 'main' works. The name convention of the initializer file doesn't matter though probably needs at least one hyphen to get picked up automatically. In the Route I get it with 'var i18n = this.container.lookup('i18n:main')'.

Today I also managed to get my YML files transpiled to JSON and funneled into a 'locales' directory AND got 'ember-cli-api-stub-static' installed with all my old api-stub data set up as new routes/[path]/index.json files AND my initial Route is now getting and setting all the View label translations.

Well, thanks for looking at it.
